[NFC] [Clang] Some core issues have changed status from tentatively ready -> ready / review

MitalAshok opened the pull request:
Also classes the "ready" status similarly to "tentatively ready" in make_cxx_dr_status

MitalAshok committed:
[NFC] [Clang] Some core issues have changed status from tentatively ready -> ready / review

Endilll commented on clang/test/CXX/drs/cwg28xx.cpp:
```diff
@@ -150,7 +150,7 @@ void g() {
 } // namespace cwg2877
 
-namespace cwg2881 { // cwg2881: 19 tentatively ready 2024-04-19
+namespace cwg2881 { // cwg2881: 19 ready 2024-04-19
```
Since this was added, wording has changed slightly, and a new ill-formed example was added. See https://github.com/cplusplus/CWG/blame/gh-pages/issues/2881.html
CC @Sirraide

Sirraide replied:
We’re already rejecting the ill-formed example in question, but we should probably add that as a test case.

Sirraide replied:
We do still need a test for template deduction, though, because from what I can tell, the wording is still the same, except that this is now also specified as causing template argument deduction to fail.

Endilll reviewed (changes requested):
Thank you for taking care of this!
But I'd like to emphasize that updating statuses of existing DR tests is not just a menial task, otherwise I wouldn't even bother making the script this strict about statuses in the first place.

MitalAshok committed:
CWG2877: Update date, update comments in example

Only the comments in the example code were changed so there should be no functional difference

https://github.com/cplusplus/CWG/commit/b7022853d731983a4b15d0a3a5b4ca49dc1ff366

MitalAshok committed:
CWG2881: Update date for new wording

A new example was also added and Clang rejects it as expected

https://github.com/cplusplus/CWG/commit/d7a3a3d302336cd467ac09867b8a19105c34346c

The wording change is:

2024-05-17:
 - Attempting to give a type not permitted for an explicit
   object parameter of a lambda's function call operator
   (7.5.5.2 expr.prim.lambda.closure)

2024-06-26:
 - Attempting to give to an explicit object parameter of
   a lambda's function call operator a type not peremitted for such
   (7.5.5.2 expr.prim.lambda.closure)

MitalAshok committed:
CWG2561: Update date

No wording changed when the status changed from tentatively ready -> ready

https://github.com/cplusplus/CWG/commit/88301b7cd100740f5aa131c405753928a071f41e#diff-7f6b9e6fa45271f479d29c851c4a8b0904f563aadff7a477b53f977113816fa1

MitalAshok committed:
CWG2819: Update date

No wording changed when the status changed from tentatively ready -> ready

https://github.com/cplusplus/CWG/commit/7d046fa9a210d8a957602dfb37405b8e083e3c66#diff-e987bc686b50066d0b5e3a9550c2d159f8e2468f1b0c678886ae0b477020740f

MitalAshok committed:
CWG2882/2883/2886: Update date

No wording changed when the status changed from tentatively ready -> ready

https://github.com/cplusplus/CWG/commit/4fb18cf3b813f9318b57e769d075a2fb9be3575e#diff-68b7577cc815e0ac6fc6d9b6dc964ef521088944b2bf3c0b9492c81241deeb09

MitalAshok committed:
CWG2885: Update date

No wording changed when the status changed from tentatively ready -> review

https://github.com/cplusplus/CWG/commit/858b5e8c1041202d8a1dc6dbd7261810df7bfc79#diff-e0e6ed8a272741d6e662722daaef5dc2682c113d6ef698b3228c3b01dfa30a93

MitalAshok committed:
run make_cxx_dr_status

MitalAshok committed:
Update cwg28xx.cpp

Use same error format as the other tests

Sirraide commented on clang/test/CXX/drs/cwg28xx.cpp:
```diff
@@ -134,23 +134,23 @@ struct A {
 
-namespace cwg2881 { // cwg2881: 19 tentatively ready 2024-04-19
+namespace cwg2881 { // cwg2881: 19 ready 2024-06-26
 
```
A test for substitution failure would be nice since that seems to be the only thing that has changed about this issue; maybe something like the following, which we seem to already be handling correctly:
```c++
template <typename L>
struct Private : private L {
    using L::operator();
    Private(L l) : L(l) {}
};

template <typename L>
constexpr auto f(L l) -> decltype(Private<L>{l}()) { return l(); }

void g() {
    int x;
    auto lambda = [x](this auto self) {};
    f(lambda); // Fails due to substitution failure here.
}
```

MitalAshok committed:
CWG2881: Add SFINAE tests

cor3ntin commented:
@MitalAshok ping

Sirraide commented on clang/test/CXX/drs/cwg28xx.cpp:
```diff
@@ -153,6 +153,8 @@ void g() {
 namespace cwg2881 { // cwg2881: 19 ready 2024-06-26
 
 #if __cplusplus >= 202302L
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Winaccessible-base"
```
This seems rather strange to me. Why are we turning off diagnostics in a test?

Endilll replied:
Yes, we shouldn't be using `#pragma clang diagnostic` in DR tests. I'll get rid of the few that are left.
If a diagnostic (ideally) should not be emitted, match it and leave a FIXME. If diagnostic is correct, you can try to change the test in a way that doesn't trigger it, but it's not required.

Sirraide reviewed (commented):
The CWG2881 tests LGTM except for one comment.

Imo it’d be better for e.g. @Endilll to approve the pr as a whole since I don’t have much experience experience w/ managing DR-related tests.

Endilll commented on clang/test/CXX/drs/cwg25xx.cpp:
```diff
@@ -139,7 +139,7 @@ struct D3 : B {
 
 #if __cplusplus >= 202302L
-namespace cwg2561 { // cwg2561: no tentatively ready 2024-03-18
+namespace cwg2561 { // cwg2561: no ready 2024-06-25
```
You should keep 2024-03-18 date, because there are no new resolution entries there.

MitalAshok replied:
It doesn't really make sense to put a date for something Clang doesn't implement yet right? So I just removed the date. This one is now status=DR on https://raw.githubusercontent.com/cplusplus/CWG/gh-pages/issues/cwg_index.html so there isn't supposed to be a date anyway

Endilll replied:
We need a date to understand what are we testing against that we reach the conclusion that we don't implement the DR

MitalAshok replied:
That makes sense, thanks for explaining. For cwg2561/cwg2883 it shouldn't need a date now because they are in the standard draft: https://eel.is/c++draft/expr.prim.lambda.closure#9.sentence-1 https://eel.is/c++draft/basic.def.odr#10.2.3

Endilll replied:
Hmm, https://cplusplus.github.io/CWG/issues/2561.html at the moment of writing says
> Section: 7.5.5.2  [[expr.prim.lambda.closure](https://wg21.link/expr.prim.lambda.closure)]     Status: ready     Submitter: Barry Revzin     Date: 2022-02-14     Liaison: EWG

while in cplusplus/CWG repo it's already DR

MitalAshok replied:
It looks like the github pages hasn't been updated for a month: https://github.com/cplusplus/CWG/deployments/github-pages but make_cxx_dr_status pulls from raw.githubusercontent.com which is always the latest. I may have been reading some stale info if that is the case

Endilll replied:
I sent Jens an email. Hopefully this will be fixed, or at least we'll get an explanation what the sync story is between pages and the repo.

Endilll replied:
> I may have been reading some stale info if that is the case

Not just you :)
I'll also recheck your changes based on the repo instead of GH pages.

Endilll replied:
Jens confirmed that GH pages are not supposed to deviate from the repo, and he's going to fix it.

Endilll commented on clang/test/CXX/drs/cwg28xx.cpp:
```diff
@@ -30,7 +30,7 @@ using U2 = decltype(&main);
 } // namespace cwg2811
 
-namespace cwg2819 { // cwg2819: 19 tentatively ready 2023-12-01
+namespace cwg2819 { // cwg2819: 19 ready 2024-06-26
```
No need to update the date here as well.

Endilll commented on clang/test/CXX/drs/cwg28xx.cpp:
```diff
@@ -216,11 +218,60 @@ void f() {
 } // namespace cwg2881
 
-namespace cwg2882 { // cwg2882: 2.7 tentatively ready 2024-05-31
+namespace cwg2882 { // cwg2882: 2.7 ready 2024-06-27
```
The date should not be updated.

Endilll commented on clang/test/CXX/drs/cwg28xx.cpp:
```diff
@@ -232,7 +283,7 @@ void f(C c) {
 } // namespace cwg2882
 
-namespace cwg2883 { // cwg2883: no tentatively ready 2024-05-31
+namespace cwg2883 { // cwg2883: no ready 2024-06-27
```
The date should not be updated.

Endilll commented on clang/test/CXX/drs/cwg28xx.cpp:
```diff
@@ -257,7 +308,7 @@ void g() {
 } // namespace cwg2883
 
-namespace cwg2885 { // cwg2885: 16 tentatively ready 2024-05-31
+namespace cwg2885 { // cwg2885: 16 review 2024-06-28
```
The date should not be updated.

Endilll commented on clang/test/CXX/drs/cwg28xx.cpp:
```diff
@@ -271,7 +322,7 @@ static_assert(!__is_trivially_constructible(B));
 } // namespace cwg2885
 
-namespace cwg2886 { // cwg2886: 9 tentatively ready 2024-05-31
+namespace cwg2886 { // cwg2886: 9 ready 2024-06-27
```
The date should not be updated.

Endilll reviewed (changes requested):
It would be nice to include here new issues that has been filed since this PR was published. CWG did not meet since St. Louis, so it shouldn't add any more work for you.

MitalAshok committed:
Merge branch 'main' into cxx_dr_status_update

MitalAshok committed:
Some statuses changed

Ready -> Accepted or DR

Treat accepted like ready

Dates were changed to the first date when there were no changes to the proposed resolution

(Only https://raw.githubusercontent.com/cplusplus/CWG/gh-pages/issues/cwg_index.html shows 'accepted' as a status, individual pages like https://cplusplus.github.io/CWG/issues/2819.html still show 'ready')

MitalAshok committed:
Remove #pragma clang diagnostic ignored

MitalAshok committed:
Merge branch 'main' into cxx_dr_status_update

Endilll reviewed (approved):
Looks good to me.
I hope that `make_cxx_dr_status` script is also happy.
Thank you for seeing this through!

MitalAshok committed:
accepted is not open

MitalAshok commented:
I've just found out "accepted" is a resolved DR, not open: https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_active.html#Issue%20Status

MitalAshok committed:
Run cxx_make_dr_status

MitalAshok commented:
@Endilll It should be good to merge after the last quick commit (which just ran cxx_make_dr_status). Could you please merge this for me? Thanks!
